File: leetcode/lc-us/cpp/1032.cpp
/*
#include <bits/stdc++.h>

using namespace std;

struct Trie {
    bool is_end;
    Trie *next[26];

    Trie() : is_end(false), next() {}

    void insert(string &s) {
        Trie *node = this;
        for (char c: s) {
            int idx = c - 'a';
            if (!node->next[idx]) node->next[idx] = new Trie();
            node = node->next[idx];
        }
        node->is_end = true;
    }
};

class StreamChecker {
    int max_length;
    string s;
    Trie *trie;
public:
    StreamChecker(vector<string> &words) {
        max_length = 0;
        trie = new Trie();
        for (string &w: words) {
            reverse(w.begin(), w.end());
            max_length = max(max_length, (int) w.size());
            trie->insert(w);
        }
    }

    bool query(char letter) {
        s.push_back(letter);
        auto node = trie;
        for (int i = (int) s.size() - 1; ~i && (int) s.size() - i <= max_length; i--) {
            int idx = s[i] - 'a';
            if (!node->next[idx]) return false;
            if (node->next[idx]->is_end) return true;
            node = node->next[idx];
        }
        return false;
    }
};
 */

#include <bits/stdc++.h>

using namespace std;

const int N = 40010;

int son[N][26], p, idx;
bool st[N];
int max_len;
string cur;

void insert(string &s) {
    p = 0;
    for (int i = (int) s.size() - 1; i >= 0; i--) {
        int u = s[i] - 'a';
        if (!son[p][u]) son[p][u] = ++idx;
        p = son[p][u];
    }
    st[p] = true;
}

bool exists() {
    p = 0;
    for (int i = (int) cur.size() - 1; i >= 0 && i >= (int) cur.size() - max_len; i--) {
        int u = cur[i] - 'a';
        if (!son[p][u]) return false;
        p = son[p][u];
        if (st[p]) return true;
    }
    return false;
}

class StreamChecker {
public:
    StreamChecker(vector<string> &words) {
        memset(son, 0, sizeof son), idx = 0, memset(st, 0, sizeof st);
        max_len = 0;
        cur.clear();
        for (auto &s: words) {
            insert(s);
            max_len = max(max_len, (int) s.size());
        }
    }

    bool query(char letter) {
        cur.push_back(letter);
        return exists();
    }
};

File: leetcode/lc-us/rust/1032.rs
use std::cmp::max;

struct Trie {
    son: Vec<Vec<usize>>,
    st: Vec<bool>,
    idx: usize,
    max_len: usize,
    cur: String,
}

impl Trie {
    fn new() -> Self {
        let son: Vec<Vec<usize>> = vec![vec![0; 26]; 40010];
        let st: Vec<bool> = vec![false; 40010];
        let idx = 0;
        let max_len: usize = 0;
        let cur: String = String::new();
        Self {
            son,
            st,
            idx,
            max_len,
            cur,
        }
    }
    fn insert(&mut self, s: &str) {
        let mut p: usize = 0;
        let bytes = s.bytes().rev();
        for c in bytes {
            let u = (c - 'a' as u8) as usize;
            if self.son[p][u] == 0 {
                self.idx += 1;
                self.son[p][u] = self.idx;
            }
            p = self.son[p][u];
        }
        self.st[p] = true;
    }
    fn exist(&mut self) -> bool {
        let mut p = 0;
        let mut i = 0;
        for byte in self.cur.bytes().rev() {
            let u = (byte - 'a' as u8) as usize;
            if self.son[p][u] == 0 {
                return false;
            }
            p = self.son[p][u] as usize;
            if self.st[p] {
                return true;
            }
            i += 1;
            if i == self.max_len {
                break;
            }
        }
        return false;
    }
}

struct StreamChecker {
    trie: Trie,
}

impl StreamChecker {
    fn new(words: Vec<String>) -> Self {
        let mut trie = Trie::new();
        for word in words.iter() {
            trie.insert(word);
            trie.max_len = max(trie.max_len, word.len());
        }
        Self {
            trie
        }
    }

    fn query(&mut self, letter: char) -> bool {
        self.trie.cur.push(letter);
        self.trie.exist()
    }
}